### PR TITLE
Modification to TID handling and new tests to cover TID and timeout

### PR DESCRIPTION
**Pull request by mike7c2**
We found an issue where sometimes receiving packets would fail following a packet timing out. This commit makes the TID handling more consistent - the rx state TID is no longer incremented when a packet is succesfully decoded/fails - previously it would be incremented to the "next expected" TID.

Tests have been added to cover various TID/timeout scenarios.

### CODE REVIEW

File: canard.c [hunk 1]
```diff
@@ -318,8 +318,7 @@ int16_t canardHandleRxFrame(CanardInstance* ins, const CanardCANFrame* frame, ui
     const bool tid_timed_out = (timestamp_usec - rx_state->timestamp_usec) > TRANSFER_TIMEOUT_USEC;
     const bool first_frame = IS_START_OF_TRANSFER(tail_byte);
     const bool not_previous_tid =
-        computeTransferIDForwardDistance((uint8_t) rx_state->transfer_id, TRANSFER_ID_FROM_TAIL_BYTE(tail_byte)) > 1;
-
+        computeTransferIDForwardDistance((uint8_t) rx_state->transfer_id, TRANSFER_ID_FROM_TAIL_BYTE(tail_byte)) >= 1;
```
Comment (pavel-kirienko) [inline canard.c]:
```suggestion
        computeTransferIDForwardDistance(TRANSFER_ID_FROM_TAIL_BYTE(tail_byte), (uint8_t) rx_state->transfer_id) > 1;
```

Context: https://github.com/UAVCAN/libcanard/issues/129#issuecomment-516951961

File: canard.c [hunk 2]
```diff
@@ -332,10 +331,15 @@ int16_t canardHandleRxFrame(CanardInstance* ins, const CanardCANFrame* frame, ui
...
-            rx_state->transfer_id++;
             return -CANARD_ERROR_RX_MISSED_START;
         }
     }
+    else if (first_frame && !not_previous_tid)
+    {
+        //This message has already been processed and hasn't timed out
+        //potential dup on bus
+        return -CANARD_ERROR_RX_DUPLICATE_TID;
+    }
```
Comment (pavel-kirienko) [inline canard.c]: This case is covered below in

```c
    if (TRANSFER_ID_FROM_TAIL_BYTE(tail_byte) != rx_state->transfer_id)
    {
        return -CANARD_ERROR_RX_UNEXPECTED_TID;
    }
```

File: canard.h [hunk 3]
```diff
@@ -79,6 +79,7 @@ extern "C" {
 #define CANARD_ERROR_RX_UNEXPECTED_TID                 15
 #define CANARD_ERROR_RX_SHORT_FRAME                    16
 #define CANARD_ERROR_RX_BAD_CRC                        17
+#define CANARD_ERROR_RX_DUPLICATE_TID                  18
```
Comment (pavel-kirienko) [inline canard.h]: This case is covered by `CANARD_ERROR_RX_UNEXPECTED_TID`

**Comment by pavel-kirienko** [comment]
@mike7c2 would you be able to address the issues I described, or should I do that?

File: canard.c (hunk 2 above)
Comment (mike7c2) [reply canard.c]: The error will be caught either way, but I think telling the user the difference between a DUPLICATE_TID and an UNEXPECTED_TID is helpful.

There are two causes I have observed for DUPLICATE_TID 
    Occurring spuriously - CAN bus duping packets
    Occurring consistently - A node calling a service on this node is not implemented correctly and is not incrementing its TID
                                             (Also related .. a node rebooting very quickly)

UNEXPECTED_TID on the other hand is indicative of packets being dropped / receiving incomplete uavcan messages.

Comment (pavel-kirienko) [reply canard.c]: I see your point, but would you not agree that the observable states are not sufficient to make a clear and unambiguous distinction between the case where CAN bus is duplicating frames and where the upper layers of the protocol are misbehaving? I am worried that the added complexity (minor complexity but still) is of no significant benefit to the application.

File: canard.c [hunk 4]
```diff
@@ -332,10 +331,15 @@ int16_t canardHandleRxFrame(CanardInstance* ins, const CanardCANFrame* frame, ui
         releaseStatePayload(ins, rx_state);
         if (!IS_START_OF_TRANSFER(tail_byte))
         {
-            rx_state->transfer_id++;
```
Comment (pavel-kirienko) [inline canard.c]: There doesn't seem to be a reason to remove this now.

File: canard.c [hunk 5]
```diff
@@ -1069,7 +1073,6 @@ CANARD_INTERNAL bool isPriorityHigher(uint32_t self, uint32_t other)
 CANARD_INTERNAL void prepareForNextTransfer(CanardRxState* state)
 {
     CANARD_ASSERT(state->buffer_blocks == NULL);
-    state->transfer_id++;
```
Comment (pavel-kirienko) [inline canard.c]: Same as above, this increment should be restored.

**Comment by pavel-kirienko** [comment]
I wonder if we could also apply these changes to the uavcan-v1.0 branch?

File: canard.c (hunk 4 above)
Comment (magshaw) [reply canard.c]: The utility of the extra reporting aside I think the changes suggested result in some bugs, and they cause some of the tests to fail:

```
/home/mike/Projects/dsdl_canard_fix/branch/libcanard/tests/test_rxerr.cpp:402: FAILED:
  REQUIRE( err == -18 )
with expansion:
  0 == -18
```

This case is making sure that we can't receive two successive packets in quick succession with the same TID. It only occurs as an issue on single frame transfers:

```
    frame.data[0] = CONSTRUCT_TAIL_BYTE(1, 1, 1, 1);
    frame.id = 0 | CANARD_CAN_FRAME_EFF;            //Set EFF
    frame.id = CONSTRUCT_SVC_ID(0, 0, 1, 20, 0);
    frame.data_len = 1;
    err = canardHandleRxFrame(&canard, &frame, 1);
    REQUIRE(err == CANARD_OK);

    //More messages with same TID fail
    err = canardHandleRxFrame(&canard, &frame, 2);
    REQUIRE(err == -CANARD_ERROR_RX_DUPLICATE_TID);
```

If we leave that case in, but still remove the `transfer_id++` calls then we get some different failures - we detect duplicate packets when a well behaved caller is incrementing its TID :

```
    //1 Frame packet, should pass
    frame.data[0] = CONSTRUCT_TAIL_BYTE(1, 1, 0, 0);
    frame.id = CONSTRUCT_SVC_ID(0, 0, 1, 20, 0);
    frame.data_len = 1;
    err = canardHandleRxFrame(&canard, &frame, 1);
    REQUIRE(CANARD_OK == err);

    //Send a start packet, should pass
    frame.data[7] = CONSTRUCT_TAIL_BYTE(1, 0, 0, 1);

    frame.id = CONSTRUCT_SVC_ID(0, 0, 1, 20, 0);
    frame.data_len = 8;                             //Data length MUST be full packet
    err = canardHandleRxFrame(&canard, &frame, 1);
    REQUIRE(CANARD_OK == err);
```

```
-------------------------------------------------------------------------------
canardHandleRxFrame no state handling, Correctness
-------------------------------------------------------------------------------
/home/mike/Projects/dsdl_canard_fix/branch/libcanard/tests/test_rxerr.cpp:201
...............................................................................

/home/mike/Projects/dsdl_canard_fix/branch/libcanard/tests/test_rxerr.cpp:242: FAILED:
  REQUIRE( 0 == err )
with expansion:
  0 == -18
```

There might be something more subtle going on, or maybe the tests aren't following the spec right but I think the PR is correct. I can make the case report the `UNEXPECTED_TID` code but I think the case and no longer incrementing the `transfer_id` need to stay

File: canard.c (hunk 2 above)
Comment (mike7c2) [reply canard.c]: The utility of the extra reporting aside I think the changes suggested result in some bugs, and they cause some of the tests to fail. Removing the case but leaving the `transfer_id++` increments we get:

```
/home/mike/Projects/dsdl_canard_fix/branch/libcanard/tests/test_rxerr.cpp:402: FAILED:
  REQUIRE( err == -18 )
with expansion:
  0 == -18
```

This case is making sure that we can't receive two successive packets in quick succession with the same TID. It only occurs as an issue on single frame transfers:

```
    frame.data[0] = CONSTRUCT_TAIL_BYTE(1, 1, 1, 1);
    frame.id = 0 | CANARD_CAN_FRAME_EFF;            //Set EFF
    frame.id = CONSTRUCT_SVC_ID(0, 0, 1, 20, 0);
    frame.data_len = 1;
    err = canardHandleRxFrame(&canard, &frame, 1);
    REQUIRE(err == CANARD_OK);

    //More messages with same TID fail
    err = canardHandleRxFrame(&canard, &frame, 2);
    REQUIRE(err == -CANARD_ERROR_RX_DUPLICATE_TID);
```

If we leave that case in, but still remove the `transfer_id++` calls then we get some different failures - we detect duplicate packets when a well behaved caller is incrementing its TID :

```
    //1 Frame packet, should pass
    frame.data[0] = CONSTRUCT_TAIL_BYTE(1, 1, 0, 0);
    frame.id = CONSTRUCT_SVC_ID(0, 0, 1, 20, 0);
    frame.data_len = 1;
    err = canardHandleRxFrame(&canard, &frame, 1);
    REQUIRE(CANARD_OK == err);

    //Send a start packet, should pass
    frame.data[7] = CONSTRUCT_TAIL_BYTE(1, 0, 0, 1);

    frame.id = CONSTRUCT_SVC_ID(0, 0, 1, 20, 0);
    frame.data_len = 8;                             //Data length MUST be full packet
    err = canardHandleRxFrame(&canard, &frame, 1);
    REQUIRE(CANARD_OK == err);
```

```
-------------------------------------------------------------------------------
canardHandleRxFrame no state handling, Correctness
-------------------------------------------------------------------------------
/home/mike/Projects/dsdl_canard_fix/branch/libcanard/tests/test_rxerr.cpp:201
...............................................................................

/home/mike/Projects/dsdl_canard_fix/branch/libcanard/tests/test_rxerr.cpp:242: FAILED:
  REQUIRE( 0 == err )
with expansion:
  0 == -18
```

There might be something more subtle going on, or maybe the tests aren't following the spec right but I think the PR is correct. I can make the case report the `UNEXPECTED_TID` code but I think the case and no longer incrementing the `transfer_id` need to stay

**Comment by pavel-kirienko** [comment]
Addressed in #142.

**Pull request closed without merging**